src/core/algorithms/od/fastod/complex_stripped_partition.h

```diff
@@ -0,0 +1,203 @@
+                             std::shared_ptr<std::vector<DataFrame::range_t>> indexes,
+                             std::shared_ptr<std::vector<size_t>> begins);
+
+public:
+    ComplexStrippedPartition() = delete;
+    ComplexStrippedPartition(ComplexStrippedPartition const& origin) = default;
+
+    ComplexStrippedPartition& operator=(ComplexStrippedPartition const& other);
+
+    std::string ToString() const;
+    void Product(short attribute);
+    bool Split(short right) const;
+
+    bool ShouldBeConvertedToStrippedPartition() const;
+    void ToStrippedPartition();
+
+private:
+    template <bool ascending>
+    bool sp_Swap(short left, short right) const {
+        for (size_t begin_pointer = 0; begin_pointer < sp_begins_->size() - 1; begin_pointer++) {
+            size_t group_begin = sp_begins_->operator[](begin_pointer);
+            size_t group_end = sp_begins_->operator[](begin_pointer + 1);
+
+            std::vector<std::pair<int, int>> values(group_end - group_begin);
+
+            for (size_t i = group_begin; i < group_end; ++i) {
+                size_t index = sp_indexes_->operator[](i);
+                values[i - group_begin] = {data_.GetValue(index, left),
+                                           data_.GetValue(index, right)};
+            }
+
+            if constexpr (ascending) {
+                std::sort(values.begin(), values.end(),
+                          [](auto const& p1, auto const& p2) { return p1.first < p2.first; });
+            } else {
+                std::sort(values.begin(), values.end(),
+                          [](auto const& p1, auto const& p2) { return p2.first < p1.first; });
+            }
+
+            size_t prev_group_max_index = 0;
+            size_t current_group_max_index = 0;
+            bool is_first_group = true;
+
+            for (size_t i = 0; i < values.size(); i++) {
+                auto const& first = values[i].first;
+                auto const& second = values[i].second;
+
+                if (i != 0 && values[i - 1].first != first) {
+                    is_first_group = false;
+                    prev_group_max_index = current_group_max_index;
+                    current_group_max_index = i;
+                } else if (values[current_group_max_index].second <= second) {
+                    current_group_max_index = i;
+                }
+
+                if (!is_first_group && values[prev_group_max_index].second > second) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    template <bool ascending>
+    bool rb_Swap(short left, short right) const {
+        for (size_t begin_pointer = 0; begin_pointer < rb_begins_->size() - 1; begin_pointer++) {
+            size_t group_begin = rb_begins_->operator[](begin_pointer);
+            size_t group_end = rb_begins_->operator[](begin_pointer + 1);
+
+            std::vector<std::pair<int, int>> values;
+            // values.reserve(...)?
```

Comment:
Why not?

Reply:
We do not know in advance how many elements there will be in the vector. We only know the number of ranges. In this case, the number of elements is much greater than the number of ranges. We can calculate the number of elements, but it will take time.